### PR TITLE
Update README.md

### PR DESCRIPTION
**Pull request by bravegrape**
## Background

Some information in the VSP README is out of date.

New work for VSP 2.0.0 means we need to update:

- the introduction
- how to run the VSP
- information on environment variables
- any outdated links

## Proposed changes

- Re-write introduction to be VSP-focused and to emphasise the new setup without an MSA
- Make information up to date and link across to tech docs where needed.
- Add new content on development mode for the VSP
- Separate information on environment variables to reflect the setup without an MSA

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,79 +2,95 @@
...
-                              # will be used during certificate rotation events (see "Generating keys for testing")
+You can run the the VSP using several commands, depending on your development needs:
 
-PORT                          # (Optional - default 50400) The TCP port where the application will listen for HTTP traffic
-LOG_LEVEL                     # (Optional - default INFO) The threshold level for logs to be written (e.g. DEBUG, INFO, WARN, or ERROR)
-```
+| Command       | Use case                               |
+| ------------- | ----------------------------------------- |
+| `development` | Local development of a VSP client <br> The VSP is connected to a test tool hosted by the GOV.UK Verify team. |
+| `server`      | End-to-end testing and running in production <br> The VSP uses the configuration provided to connect to <br> a full-scale of the GOV.UK Verify Hub.                                          |
```
Comment (chrisholmes) [inline README.md]: full-scale version?

File: README.md [hunk 2]
```diff
@@ -2,79 +2,95 @@
...
+| `check`       | Validates your configuration file.        |
 
-As Verify Service Provider is a Dropwizard application, you can also configure it with all [options provided by Dropwizard](http://www.dropwizard.io/1.3.5/docs/manual/configuration.html).
+### `development`
 
-### Generate keys for testing
+You can use `development` command if you're building your own client for the VSP.The command starts the VSP connected to a test tool hosted by the GOV.UK Verify team. The test tool acts as a placeholder for the GOV.UK Verify Hub. This means you can use your local setup to test if your service can respond appropriately to all possible scenarios in a Verify journey.
```
Comment (chrisholmes) [inline README.md]: You can use the?

File: README.md [hunk 3]
```diff
@@ -2,79 +2,95 @@
...
-In order to generate keys for testing, we recommend using [OpenSSL](https://www.openssl.org).
+When running the VSP using the `development` command, it initialises the test tool by:
 
-You can generate a private key by:
-1. generating an RSA key in PEM format
-2. converting the key to base64 encoded PKCS8
+- generating its self-signed keys and certificates
+- adding the keys and certificates to the VSP configuration
+- setting the test tool environment in the VSP configuration
+- initialising an instance of the test tool
```
Comment (chrisholmes) [inline README.md]: > initialising an instance of the test tool
This feels a bit awkward to me. Could we say something like: 'initialising a testing session with the test tool'?

File: README.md [hunk 4]
```diff
@@ -88,17 +104,42 @@ You can check the application is running by calling the healthcheck endpoint wit
...
+
+| Variable | Description |
+| -------- | ----------- |
+| `VERIFY_ENVIRONMENT` | The GOV.UK Verify Hub environment to run in.<br/>For example `PRODUCTION`, `INTEGRATION`|
+| `SERVICE_ENTITY_IDS` | A JSON string array with the service's entity ID, for example `'["http://entity-id"]'`. If you have several services using one VSP deployment,<br/> the array should contain all of their service entity IDs. |
+| `SAML_SIGNING_KEY` | A base64 encoded RSA private key used for signing the request to GOV.UK Verify Hub.|
+| `SAML_PRIMARY_ENCRYPTION_KEY`| A primary base64 encoded PKCS8 RSA private key used to decrypt SAML responses.|
+|`SAML_SECONDARY_ENCRYPTION_KEY`|(Optional - default empty) A secondary base64 encoded PKCS8 RSA private key is used to decrypt SAML responses. This parameter applies during [key rotation][key-rotation] events.|
+| `PORT` | (Optional - default `50400`) The TCP port where the application will listen for HTTP traffic|
+| `LOG_LEVEL` | Optional - default `INFO`) The threshold level for logs to be written, for example `DEBUG`, `INFO`, `WARN`, or `ERROR`) |
```
Comment (chrisholmes) [inline README.md]: We need to add something about eIDAS here, but we can do it as part of another commit.

File: README.md [hunk 5]
```diff
@@ -88,17 +104,42 @@ You can check the application is running by calling the healthcheck endpoint wit
...
+
+If you are using the legacy version involving a [Matching Service Adapter (MSA)](https://github.com/alphagov/verify-matching-service-adapter), two additional environment variables apply:
+
+| Variable         | Description                         |
+| ---------------- | ----------------------------------- |
+| MSA_ENTITY_ID    | The `entityId` of the service's MSA |
```
Comment (chrisholmes) [inline README.md]: They probably need to edit the YAML file too so those environment variables can be used.

File: README.md (hunk 2 above)
Comment (bravegrape) [reply README.md]: 😅

Comment (chrisholmes) [reply README.md]: __WILDCARD__: Maybe we want to call it test service rather than tool as tool to me implies that you are running something locally.

File: README.md [hunk 6]
```diff
@@ -129,4 +170,12 @@ If you think you have discovered a security issue in this code please email [dis
 For non-security related bugs and feature requests please [raise an issue](https://github.com/alphagov/verify-service-provider/issues/new) in the GitHub issue tracker.
 
 ## Licensing
-[MIT License](https://github.com/alphagov/verify-service-provider/blob/master/LICENSE)
+[MIT License][mit-license]
+
+[key-rotation]: https://www.docs.verify.service.gov.uk/maintain-your-connection/rotate-keys/
+[vsp-get-started]: https://www.docs.verify.service.gov.uk/get-started-with-vsp/
+[vsp-api]: https://github.com/alphagov/verify-service-provider/blob/master/architecture-decisions/verify-service-provider-api.swagger.yml
```
Comment (chrisholmes) [inline README.md]: This is going to be a little wrong, but we can fix it later.

File: README.md [hunk 7]
```diff
@@ -2,79 +2,95 @@
...
+Using the VSP removes the need for services to handle SAML by:
 
-You will need to host Verify Service Provider on your own infrastructure.
+* generating SAML requests to send to the GOV.UK Verify Hub
+* translating SAML responses from the GOV.UK Verify Hub into JSON
 
-Using Verify Service Provider is just one part of connecting to GOV.UK Verify. Refer to the [technical onboarding guide](https://www.docs.verify.service.gov.uk) for more information about connecting to GOV.UK Verify.
+Services will need to host the VSP on their own infrastructure.
+
+The VSP allows connected services to handle signing and encryption key rotation without service downtime.
```
Comment (chrisholmes) [inline README.md]: This feels a little out of place now. Maybe we could have it inside a subheading?

File: README.md (hunk 3 above)
Comment (bravegrape) [reply README.md]: Yes, we can. I think someone phrased it like that ages ago and it stuck around 🤷‍♀️

File: README.md (hunk 4 above)
Comment (bravegrape) [reply README.md]: 👍 Let me know if you want me to write a separate card for this.

File: README.md (hunk 6 above)
Comment (bravegrape) [reply README.md]: Do you mean the VSP API reference location? Yeah, we should remember to include the README updates when we pick up the bit of work...

File: README.md (hunk 2 above)
Comment (bravegrape) [reply README.md]: I like the idea. 'Test tool' evolved from 'Compliance Tool' anyway, not much thought went into the right way to call it.

How about 'testing service'?
Does a service have commands or modes? 😏

File: README.md (hunk 1 above)
Comment (bravegrape) [reply README.md]: Curses. I think I accidentally deleted 'deployment'. I'm changing to 'full-scale deployment' 👍

**Comment by bravegrape** [comment]
[Re-write README.md for the VSP](https://trello.com/c/VYIwbgIC/16-re-write-readmemd-for-the-vsp)

File: README.md [hunk 8]
```diff
@@ -2,79 +2,98 @@
...
 
-### Run
+For more information on building your own client using the `development` command, see the [technical documentation on how to get started][vsp-get-started].
 
-To run the application, export your environment variables and start the application with:
+### `server`
+
+Use the `server` command when running the VSP in an environment containing a full-scale deployment of the GOV.UK Verify Hub, for example the Integration or Production environments.
+
+For example, to run the VSP using the environment and security configuration in `verify-service-provider.yml`, export your environment variables and run:
```
Comment (alan-gds) [inline README.md]: Do we need any elaboration here, or are we confident the RP devs will understand they need to look inside the config file to find out which environment variables they need to export?

Comment (bravegrape) [reply README.md]: Could link to the 'Configuration' section which talks about environment variables?

File: README.md (hunk 5 above)
Comment (chrisholmes) [reply README.md]: This comment still applies. I think we should remove this table and leave the sentence on line 142, but remove the words: environment variables.

Comment (bravegrape) [reply README.md]: Fair enough. I'll remove the table since they won't be able to use the information in the table anyway.  I've pushed a commit with this change 👌
